scheduler/src/cook/mesos/util.clj

```diff
@@ -552,3 +552,28 @@
   [ch ch-size]
   (->> (repeatedly ch-size #(async/poll! ch))
        (remove nil?)))
+
+(defn xform-pipe
+  "Reads elements from the `in-chan` channel and supplies elements to the `out-chan` channel.
+   Maintains a state (initialized to `initial-state`) that is updated by applying the reducing function
+   `in-xform` to the current state and the incoming element `(in-xform state element)`.
+   `out-xform` accepts the current state (i.e. `(out-xform state)`) and produces `[reset-state out-value]`.
+   When the `out-chan` channel can receive an element, the `value` is put on the channel and the state is
+   set to the `reset-state` returned from the `out-xform`. Invocation of `out-xform` does not guarantee that
+   `out-value` will be consumed, hence it should not have side-effects. The optional parameter `on-consumed`
+   can be used to register notifications of when a value has been supplied to the `out-chan` channel.
+
+   Note: This function does not perform error handling, exceptions must be explicitly handled in the
+   provided functions (i.e. in-xform, out-xform, on-consumed and on-finished)."
+  [in-chan initial-state in-xform out-chan out-xform &
+   {:keys [on-consumed on-finished] :or {on-consumed identity, on-finished #()}}]
+  (async/go-loop [state initial-state]
+    (let [[reset-state out-value] (out-xform state)
```

Comment:
Holy shit, the [alts code](https://github.com/clojure/core.async/blob/master/src/main/clojure/clojure/core/async.clj#L237) is crazy. I don't understand how they are achieving the promise that it will block until one is available. It looks like they are trying each channel and returning if none are ready.